### PR TITLE
[AIRFLOW-2336] Use hmsclient instead of hive-thrift-py

### PR DESCRIPTION
**Pull request by gglanzani**
The package hmsclient is Python2/3 compatible and offer a handy context
manager to handle opening and closing connections.

Make sure you have checked _all_ steps below.

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issue and references them in the PR title
    - https://issues.apache.org/jira/browse/AIRFLOW-2336

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR allows the hive_hook to be compatible with Python 3 by using the `hmsclient` package instead of `hive-thrift-py`.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

It replaces existing functionality, so existing tests should cover it.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by Fokko** [comment]
The failed test is a bit flakey:
```
======================================================================
19) ERROR: test_authorized (tests.www.api.experimental.test_password_endpoints.ApiPasswordTests)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/www/api/experimental/test_password_endpoints.py line 64 in test_authorized
      headers={'Authorization': 'Basic aGVsbG86d29ybGQ='}  # hello:world
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/werkzeug/test.py line 840 in post
      return self.open(*args, **kw)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/testing.py line 127 in open
      follow_redirects=follow_redirects)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/werkzeug/test.py line 803 in open
      response = self.run_wsgi_app(environ, buffered=buffered)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/werkzeug/test.py line 716 in run_wsgi_app
      rv = run_wsgi_app(self.application, environ, buffered=buffered)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/werkzeug/test.py line 923 in run_wsgi_app
      app_rv = app(environ, start_response)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1997 in __call__
      return self.wsgi_app(environ, start_response)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1985 in wsgi_app
      response = self.handle_exception(e)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1540 in handle_exception
      reraise(exc_type, exc_value, tb)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/_compat.py line 33 in reraise
      raise value
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1982 in wsgi_app
      response = self.full_dispatch_request()
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1614 in full_dispatch_request
      rv = self.handle_user_exception(e)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1517 in handle_user_exception
      reraise(exc_type, exc_value, tb)
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/_compat.py line 33 in reraise
      raise value
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1612 in full_dispatch_request
      rv = self.dispatch_request()
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/flask/app.py line 1598 in dispatch_request
      return self.view_functions[rule.endpoint](**req.view_args)
    airflow/contrib/auth/backends/password_auth.py line 203 in decorated
      response = function(*args, **kwargs)
    airflow/www/api/experimental/endpoints.py line 78 in trigger_dag
      dr = trigger.trigger_dag(dag_id, run_id, conf, execution_date)
    airflow/api/common/experimental/trigger_dag.py line 61 in trigger_dag
      external_trigger=True
    airflow/utils/db.py line 74 in wrapper
      return func(*args, **kwargs)
    airflow/models.py line 4006 in create_dagrun
      run.refresh_from_db()
    airflow/utils/db.py line 74 in wrapper
      return func(*args, **kwargs)
    airflow/models.py line 4706 in refresh_from_db
      DR.run_id == self.run_id
    .tox/py35-backend_sqlite/lib/python3.5/site-packages/sqlalchemy/orm/query.py line 2820 in one
      raise orm_exc.NoResultFound("No row was found for one()")
   NoResultFound: No row was found for one()
```

**Comment by Fokko** [comment]
@mistercrunch Any thoughts on this? Do you know if this is still being used?  The package is Deprecated according to the author: https://github.com/youngwookim/hive-thrift-py

Cheers!

**Comment by bolkedebruin** [comment]
@saguziel @aoen will merge if you are not objecting. The previous package is deprecated so we need to move anyway.

I see you are the author of hmsclient @gglanzani is it based on the earlier client? Or completely new? How much exposure has it?

**Comment by gglanzani** [comment]
It has been generated using the latest thrift files (see file generate.py in the repo, or something like that). The hmsclient class is a thin wrapper with a couple of useful methods (such as enter, exit, etc).

What do you mean by exposure though?

**Comment by bolkedebruin** [comment]
@gglanzani I like the update, bit it needs tests. The hive-thrift-py has been in for a long long time and its now being replaced by something new and shiny. That's good but then tests are required. You have a bit of time before the RCs of 1.10. We do betas first.

btw does your implementation support Kerberos yet? ;-)

**Comment by gglanzani** [comment]
@bolkedebruin tests in hmsclient or Airflow? Is the metastore available in a Docker image in Airflow, or should I use something external?

**Comment by Fokko** [comment]
Could you maybe bump the test coverage: https://codecov.io/gh/apache/incubator-airflow/src/1d3bb5470711a935b36b6a0ab4c7ec414d460d75/airflow/hooks/hive_hooks.py ?

**Comment by bolkedebruin** [comment]
Hive is already running and a DB is also available. So it is probably just an issue of making the inegration work?

**Comment by gglanzani** [comment]
@fokko the thing is, there's no metastore available in the tests (in form of some Docker image) is there?

**Comment by Fokko** [comment]
Yes, Bolke is right. We're using `apache-hive-1.2.1.2.3.2.0-2950`: https://github.com/apache/incubator-airflow/blob/c7a472ed6b0d8a4720f57ba1140c8cf665757167/scripts/ci/setup_env.sh#L97

**Comment by gglanzani** [comment]
Ok, let's see how bad I am at writing tests blind (no CI locally :)

**Comment by artwr** [comment]
@gglanzani You can setup Travis CI on your fork for free as it is open source. If you setup to build your branch you can check your results there faster, as your build will be in its own queue.

**Comment by gglanzani** [comment]
@bolkedebruin I think I've fixed the last test. Let me know if those were the sort of tests you were thinking about.

BTW, is it normal that everything is empty in the warehouse when testing? By looking at [these] tests I got the impression that the Airflow db was meant to be there (but it's not).

Also: what do you think about upgrading the Hive version? The current one is ridded with pesky bugs, such as [this] which I encountered while writing the tests.

@artwr Thanks, but I'd have to wait at least 20 minutes anyway. I would love to have a container somewhere that speeds up testing cases such as this (not saying it's easy or that you folks should do that).

[these]: https://github.com/apache/incubator-airflow/blob/c7a472ed6b0d8a4720f57ba1140c8cf665757167/tests/operators/hive_operator.py#L46
[this]: https://issues.apache.org/jira/browse/HIVE-12200

**Comment by bolkedebruin** [comment]
@gglanzani that would be awesome! Don’t know about the warehouse honestly, getting hive up and running in the container is pretty tough (no hdfs) so I’m not sure if it is entirely ‘normal’. On upgrading hive I’m a little in doubt. How widespread is the newer hive? It really takes a long time before everyone migrated. But given the fact our tests are not testing all of hive (and they should have to), I think it is fine to upgrade to a newer version. If you can get it up and running of course. Still in a separate PR please.

**Comment by bolkedebruin** [comment]
Ideally we split integration tests away from the unit tests. That should be faster overall. Quite an undertaking though.

**Comment by bolkedebruin** [comment]
Or have a gitlab ci version available.

**Comment by gglanzani** [comment]
@bolkedebruin 
So the Python 3 tests seem to be going fine, but 2.7 gives me a timezone error. Is there something I should be aware of? https://travis-ci.org/apache/incubator-airflow/jobs/370751659

@Fokko Do you see something in the test code that clearly violates something in 2.7 and that could explain the failing tests? I'm using `DEFAULT_DATE = datetime.datetime(2015, 1, 1)` as start and end date, which is timezone unaware.

File: tests/hooks/test_hive_hook.py [hunk 1]
```diff
@@ -18,13 +18,68 @@
 # under the License.
 #
 
+import datetime
```
Comment (bolkedebruin) [inline tests/hooks/test_hive_hook.py:21]: Use airflow.utils.timezone you are calling ti.run directly, you can’t do that with naive date times.

**Review by bolkedebruin — CHANGES_REQUESTED** [review]

File: tests/hooks/test_hive_hook.py (hunk 1 above)
Comment (gglanzani) [reply tests/hooks/test_hive_hook.py:21]: Thanks, pushed!

**Comment by gglanzani** [comment]
@bolkedebruin everything should be good to go now

**Review by Fokko — APPROVED** [review]
@gglanzani Awesome work on the test!

**Comment by bolkedebruin** [comment]
@gglanzani nice one!!

**Pull request closed without merging**